Gramine single-stepping example

aegis2501 opened the pull request.

jovanbulck commented on sdk/gramine/0001-SGX-Step-Gramine-patches-to-reconfigure-AEP-TCS.patch:
```diff
@@ -1,22 +1,27 @@
+     endif
+ 
++  # XXX taken from: https://stackoverflow.com/questions/34663124/link-static-library-to-shared-library-or-to-a-binary
++  libsgxstepdir = '<ENTER ABSOLUTE PATH OF LIBSGXSTEP HERE>'
```
not sure, why is the manual absolute path needed here?

For me, I could compile and link gramine and libsgxstep with my earlier [patch](https://github.com/jovanbulck/sgx-step/blob/master/sdk/gramine/0001-SGX-Step-Gramine-patches-to-reconfigure-AEP-TCS.patch).

If we can avoid this manual entering of the absolute path twice here, that would certainly be easier :)

aegis2501 replied:
If I remember correctly it would compile but then have problems finding the functions from the library later on (it's been a while though)

jovanbulck commented on sdk/gramine/0002-helloworld-with-Nemesis-in-Gramine.patch:
```diff
@@ -0,0 +1,269 @@
+@@ -6,6 +6,185 @@
+ #include "pal_ecall_types.h"
+ #include "pal_rpc_queue.h"
+ 
```
Thanks for the code! Ideally, for future maintenance, this chunk of code that does the actual single-stepping would be in a separate C file in the `sdk/gramine` directory, so that it would have proper version control etc. 

That is: the patches should ideally be only the smallest changes needed to link into SGX-Step and then the actual attack logic can be maintained and version controlled as normal C files. This has the advantage that the attack code is more easy to read and understand and extend, and, most importantly, the changes to the target runtime are minimal and can hopefully be ported easily across Gramine versions :)

Concretely I suggest:
1. moving the single-stepping logic from `host_ecalls.c` to `sdk/gramine/aep.c` or so and only keep in the patch file the minimal setup before the first ecall
2. changing the gramine build system in the patch to also include this `../aep.c` file (or alernatively, it should be possible to link it into libsgxstep instead)

(Also related, in the future, I'd like to properly refactor the single-stepping abstraction issue #2  such that all the code duplication with `app/bench/main.c` is avoided )

jovanbulck reviewed (commented):
Thanks a lot for the code @aegis2501 and sorry for the delay due to the holiday season :christmas_tree:

I'm sure this is already a great help in case ppl want to further explore Gramine :)

I didn't test the patches myself (yet), but I did a code review and left some comment for possible further improvement. My main comment before merging this upstream would be to reduce the size of the patches as much as possible (with the eye on future usability and compatibility) as patches are notoriously unreadable and hard to maintain.

Specifically, I suggest to (1) parameterize libsgxstep to avoid patches there; and (2) move the sizable attack logic to a separate C file in the `sdk/gramine` directory.

Let me know if you think these changes are feasible?

aegis2501 commented on sdk/gramine/0000-Libsgxstep-patches.patch:
```diff
@@ -0,0 +1,43 @@
+ 
+ extern void *apic_base;
+-void *dummy_pt = NULL;
++//void *dummy_pt = NULL;
```
Yes, this can be removed from the patch.

aegis2501 commented on sdk/gramine/0000-Libsgxstep-patches.patch:
```diff
@@ -0,0 +1,43 @@
+ 
+     addr = (volatile uint32_t *)(apic_base + reg);
+-    asm volatile ("movl %1, %0\n\t"
++    __asm__ volatile ("movl %1, %0\n\t"
```
The old syntax did indeed not compile for gramine.

aegis2501 commented on sdk/gramine/0000-Libsgxstep-patches.patch:
```diff
@@ -0,0 +1,43 @@
+ //TODO determine this at runtime..
+ // SSA framesize for Gramine seems to be as follows
+-// #define SGX_SSAFRAMESIZE            16384
++#define SGX_SSAFRAMESIZE            16384
```
Sure, that's probably the more elegant method anyways :)

aegis2501 commented on sdk/gramine/patch_sgxstep.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
```
Yes :)

jovanbulck commented:
Hi aegis,

Thank you for making the changes!  I am busy this week, but I'll properly review and test this PR end of this week or next week :)

One thing that should def still be done before merging this: can you move the file `libsgxstep/aep.c` to `sdk/gramine/aep.c`?

At least for now this is still gramine-specific and should, therefore, be in the gramine directory (you should be able to compile it separately and link it into gramine similar to libsgxstep).

In near future, I want to finally properly refactor all the duplicated single-stepping logic into a libsgxstep abstraction (#2 ) But until then this should reside in sdk/gramine.

aegis2501 commented:
Hello again, apologies for the late reply! I've moved `libsgxstep/aep.c` to `sdk/gramine` (also had to adjust the Makefile in libsgxstep a bit). I also fixed the patches as they appeared to be misaligned in some sections for whatever reason, but they should work with the used `gramine` commit out of the box.
